### PR TITLE
fix(search): shared search_tokens lookup + stop silently ILIKE-ing encrypted columns (#2990)

### PR DESCRIPTION
**Pull request by marcinwadon**
## Summary

First slice of #2990 (the `$ilike` → `query-engine`/`search_tokens` migration). It does **not** attempt the ~62-file sweep — it lands the missing prerequisite, closes the silent-failure mode that makes this class of bug invisible, and fixes the two routes where the defect is already provable today.

Scope was agreed as "PR A" of three:

- **PR A (this one)** — shared helper, loud fallback, the two confirmed-broken routes.
- **PR B** — the `AGENTS.md` avoid-clause (AC 3) and the extended `$ilike` regression guard (AC 4).
- **PR C+** — the long tail, per module, behind a spec (see *Why not the full sweep* below).

## What changed

**1. One shared `search_tokens` lookup** (`@open-mercato/shared/lib/search/tokenLookup`)

`customers`, `auth` and `messages` each carried a hand-rolled Kysely lookup against `search_tokens` with subtly different tenant/organization scoping. Migrating 25 more call sites on top of that would have produced 28 copies. All three now delegate to one helper.

The helper returns a discriminated result:

```ts
{ matched: true, ids: string[] }                                   // the index answered
{ matched: false, reason: 'empty-query'|'search-disabled'|'no-tokens' }  // never consulted
```

That distinction is exactly what the previous `string[] | null` shape lost, and it is what every future migration needs to decide whether it may drop the caller's own predicate. The three existing call sites keep their old shape via `findEntityIdsBySearchTokensCompat`, so this part is behavior-preserving.

**2. The silent fallback is now loud** (`lib/query/ciphertext-search-warning`)

Both query engines already route `like`/`ilike` through `search_tokens` when tokens exist — and fall back to a raw column predicate when they do not. On an encrypted column that fallback compares a plaintext pattern against ciphertext: it matches nothing, and the endpoint returns an empty page indistinguishable from a genuine no-result. The only trace was a debug line behind `OM_SEARCH_DEBUG`.

This is arguably worse than the bug in the issue, because it hides it. The engines now `logger.warn` with the entity, the field and why tokens were unavailable. Cost control: the check runs **only** on the fallback path, **only** when encryption is enabled, and warns **once per entity/tenant/field per process**.

**3. The two provably broken routes**

Both read through `findAndCountWithDecryption` — so the author knew the columns are encrypted — yet filtered them with `$ilike`:

| Route | Encrypted columns filtered with `$ilike` |
|---|---|
| `checkout/api/transactions` | `email`, `first_name`, `last_name` |
| `inbox_ops/api/proposals` | `summary` |

The token match is **unioned into** the existing predicate rather than replacing it: where tokens exist the encrypted columns become searchable again; where they do not, the request behaves exactly as before. **No entity needs reindexing for this to be safe to ship.**

**4. Drive-by, same lines:** checkout also filtered `{ id: { $ilike } }` against a `uuid` column. Postgres has no `uuid ~~* text` operator, so *any* search term on that endpoint raised 42883 and 500'd. A term that parses as a UUID now matches the id exactly; otherwise the clause is omitted.

## Why not the full sweep

Two things AC 2 does not resolve, which is why the remaining ~25 raw-ORM sites need a spec first:

- **Token search is not substring search.** `OM_SEARCH_MIN_LEN=3`, expansion is prefix-only. `ilike '%ohn%'` finds "John" today; tokens will not. `currencies` searches `symbol` (`$`, `zł`) and 2-letter codes — below the tokenizer floor. A blanket migration is a regression on non-encrypted, short-token fields, so it needs a per-field decision (encrypted → tokens; codes/symbols → keep `ilike` or add a deterministic hash column).
- **Several affected entities are not in `query_index` at all.** The generic indexer only picks up entities that have custom-field definitions (`query_index/di.ts`), so for those, "migrate the filter" is really "index the entity + backfill + reindex" — a different order of magnitude.

## Verification

Full CI-mirroring gate from `.ai/agentic.config.json`, run locally:

| Command | Result |
|---|---|
| `yarn build:packages` | pass |
| `yarn generate` | pass |
| `yarn i18n:check-sync` | pass — all locales in sync |
| `yarn i18n:check-usage` | pass (advisory) |
| `yarn typecheck` | pass — 21/21 |
| `yarn test` | pass — 22/22 packages¹ |
| `yarn build:app` | pass |

¹ `create-mercato-app` has 4 failing suites (`agent-harness-evaluator`, `*-oracles`). Verified pre-existing: they fail identically on a clean `origin/develop` with this branch stashed.

New coverage: 16 cases for the lookup helper (scope shapes, skip reasons, all-tokens matching), 11 for the warning (encrypted vs not, camelCase↔snake_case, warn-once, never-throws), 6 for the checkout route, 2 added to inbox_ops.

## Reviewer notes

- The `search-disabled` reason is derived from `searchConfig.enabled` rather than the engines' `searchEnabled`, because the latter also folds in the missing-table and `omitAutomaticTenantOrgScope` cases — those are "no usable tokens", not "the operator switched search off".
- The warn cache is process-global and capped at 5000 entries (cleared wholesale on overflow); `resetCiphertextLikeWarnCache()` is exported as a test seam.
- `findEntityIdsBySearchTokens` uses `field = x` for a single field and `field IN (…)` for several — equivalent SQL, kept so the existing auth-route assertions stay untouched.

Refs #2990. Related: #2946, #2932.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by pkarw** [comment]
Thanks for taking care of this one!

**Comment by pkarw** [comment]
🤖 `om-auto-review-pr` started by @pkarw at 2026-08-01T07:48:32Z. Other auto-skills will skip this PR until the lock is released.

**Comment by pkarw** [comment]
Label set to `review` because this PR is ready for code review.

File: packages/checkout/src/modules/checkout/api/transactions/route.ts [hunk 1]
```diff
@@ -27,12 +31,32 @@ export async function GET(req: Request) {
...
+      // stores hashes of the plaintext. Issue #2990.
+      const searchOr: Record<string, unknown>[] = [
+        { email: { $ilike: pattern } },
+        { firstName: { $ilike: pattern } },
+        { lastName: { $ilike: pattern } },
+      ]
+      // `id` is a uuid column: ILIKE against it has no Postgres operator and
+      // raises 42883, so only an exact id lookup is meaningful here.
+      if (UUID_PATTERN.test(search)) searchOr.push({ id: search })
+      const tokenMatch = await findEntityIdsBySearchTokens({
```
Comment (pkarw) [inline packages/checkout/src/modules/checkout/api/transactions/route.ts:48]: **Major:** This lookup has no incremental producer for `checkout:checkout_transaction`. Creation emits `checkout.transaction.created`, but the generic query-index listener derives/listens for `checkout.checkout_transaction.created` (and is custom-field-gated); checkout has no explicit `query_index.upsert_one` bridge. New transactions therefore have no token rows, so encrypted email/name search still falls through to ciphertext `ILIKE` and returns empty. Please wire a scoped reindex subscriber and cover the real create → index → encrypted list-search path.

File: packages/core/src/modules/inbox_ops/api/proposals/route.ts [hunk 2]
```diff
@@ -47,7 +49,24 @@ export async function GET(req: Request) {
...
     if (query.search) {
-      where.summary = { $ilike: `%${escapeLikePattern(query.search)}%` }
+      // `summary` is covered by the inbox_ops encryption map, so an ILIKE alone
+      // compares the pattern against ciphertext and matches nothing once
+      // encryption is on. Union it with the token index, which stores hashes of
+      // the plaintext. Issue #2990.
+      const searchOr: FilterQuery<InboxProposal>[] = [
+        { summary: { $ilike: `%${escapeLikePattern(query.search)}%` } },
+      ]
+      const tokenMatch = await findEntityIdsBySearchTokens({
```
Comment (pkarw) [inline packages/core/src/modules/inbox_ops/api/proposals/route.ts:59]: **Major:** The same lifecycle gap exists here. Proposals emit `inbox_ops.proposal.created`, while the canonical entity ID is `inbox_ops:inbox_proposal`; no explicit bridge emits `query_index.upsert_one`. The mocked route test cannot catch that the real token table is empty. Please add the scoped bridge and a real encrypted-search integration test, including a cross-tenant/org negative assertion.

**Review by pkarw — CHANGES_REQUESTED** [review]
# Code Review: fix(search): shared search_tokens lookup + stop silently ILIKE-ing encrypted columns (#2990)

## Summary

This adds a useful shared token lookup, replaces several duplicated lookups, and attempts to repair encrypted-field filtering in checkout transactions and inbox proposals. The lookup itself preserves tenant/organization scope, but both repaired routes depend on token rows that their entities do not incrementally produce, so the principal bug remains on realistic newly-created records.

## Verdict

request changes — one major correctness defect and missing end-to-end regression coverage make the fix ineffective on its two new API paths; this repository also treats the Medium findings below as blocking.

## Validation Gate

Validation source: GitHub checks for head `f4af5fcc0a77475a13d420fd2fabd262f545a150` (all reported checks successful; required-check metadata endpoint returned 404, so every reported check was treated as required). No local fallback was run because the CI signals were usable.

| Command | Status | Notes |
|---------|--------|-------|
| `yarn build:packages` | PASS | GitHub `prepare` |
| `yarn generate` | PASS | GitHub `prepare` |
| `yarn build:packages` | PASS | GitHub `prepare` rebuild |
| `yarn i18n:check-sync` | PASS | GitHub `test` |
| `yarn i18n:check-usage` | PASS | GitHub `test` |
| `yarn typecheck` | PASS | GitHub `test` affected-scope typecheck |
| `yarn test` | PASS | GitHub `test` affected-scope tests |
| `yarn build:app` | PASS | GitHub `prepare`; all 15 ephemeral-integration shards also passed |

## Findings

### Major

1. **The two newly repaired list routes have no incremental token producer.**  
   Files: `packages/checkout/src/modules/checkout/api/transactions/route.ts:48`, `packages/core/src/modules/inbox_ops/api/proposals/route.ts:59`, `packages/checkout/src/modules/checkout/commands/transactions.ts:112`, `packages/core/src/modules/inbox_ops/subscribers/extractionWorker.ts:582`.

   Checkout emits `checkout.transaction.created` for entity `transaction`, while the canonical entity ID is `checkout:checkout_transaction`; inbox emits `inbox_ops.proposal.created`, while its canonical ID is `inbox_ops:inbox_proposal`. The generic query-index bridge therefore listens for different event names, and its handler is custom-field-gated anyway. Neither module has an explicit `query_index.upsert_one` bridge for these events. Newly-created records consequently have no `search_tokens`, so the added ID union contributes nothing and the remaining `ILIKE` still compares plaintext to ciphertext. Add explicit, tenant-scoped reindex subscribers (and an existing-data reindex path if required) for both canonical entity IDs.

2. **The bug fix lacks integration coverage for the behavior that failed.**  
   Files: `packages/checkout/src/modules/checkout/api/transactions/__tests__/route.test.ts:1`, `packages/core/src/modules/inbox_ops/api/proposals/__tests__/route.test.ts:1`.

   Both route tests mock `findEntityIdsBySearchTokens` to return IDs, so they cannot detect the missing indexing lifecycle above. The repository requires self-contained integration coverage for affected API paths, particularly encryption-, tenant-, and payment-sensitive behavior. Add tests that create a transaction/proposal, run the real index path under encryption, search the list API by encrypted text, assert the right record is returned, and assert another tenant/organization cannot match it.

### Minor

1. **Ciphertext fallback warnings miss real silent fallbacks.**  
   Files: `packages/shared/src/lib/query/engine.ts:337`, `packages/core/src/modules/query_index/lib/engine.ts:450`, `packages/shared/src/lib/query/ciphertext-search-warning.ts:73`.

   The engines warn only when token search is globally inactive. When an entity has token rows but a specific value tokenizes to zero hashes (for example, a term below `OM_SEARCH_MIN_LEN`), the code still falls through to raw `LIKE/ILIKE` without warning. BasicQueryEngine also excludes join filters from the warning set. In addition, the warn-once cache records unencrypted candidates before checking the encryption map, which can suppress a later legitimate warning if encryption configuration changes. Detect the actual fallback cases, include joined encrypted fields with their owning entity, and cache only warnings that were emitted; add engine-level tests.

2. **The new public helper and call sites introduce prohibited `any` types.**  
   Files: `packages/shared/src/lib/search/tokenLookup.ts:5`, `packages/shared/src/lib/query/ciphertext-search-warning.ts:11`, `packages/checkout/src/modules/checkout/api/transactions/route.ts:49`, `packages/core/src/modules/inbox_ops/api/proposals/route.ts:60`.

   Define a precise Kysely database shape for `search_tokens`, use the shared `EntityId` type for encryption service calls, and remove the new `getKysely<any>()` / builder `any` uses (including the newly documented example).

3. **The documented cross-package import is missing an explicit package export.**  
   Files: `packages/shared/package.json:13`, `packages/shared/AGENTS.md:55`.

   The wildcard currently resolves `@open-mercato/shared/lib/search/tokenLookup`, but repository guidance requires an explicit export entry for every new shared utility intended for cross-package consumption. Add a typed/default export mapping for this subpath.

4. **The new additive contract and query behavior are not reflected in the active search spec.**  
   File: `.ai/specs/2026-07-24-search-architecture-clarification-and-evolution.md`.

   The PR adds a documented public import and alters direct-list search behavior, but the active architecture spec still describes the pre-change state and the PR does not record migration/backward-compatibility implications. Update the existing spec's relevant architecture, coverage, and compatibility sections.

## Breaking Changes

- [x] No exported/public symbol removed or renamed without a deprecation path
- [x] No function signature changed in a breaking way
- [x] No required type field removed or narrowed
- [x] No HTTP route URL or method removed/renamed
- [x] No response field removed or retyped
- [x] No event/message name or payload field removed
- [x] No CLI surface changed
- [x] No database schema changed
- [x] No config key renamed or default silently changed
- [ ] Additive public surface is implemented, but the explicit export and spec/compatibility record are incomplete as noted above

## Test Coverage

The shared lookup and warning helpers have useful unit tests, the refactored call sites retain unit coverage, and CI passed. Missing: producer-to-`search_tokens` lifecycle tests, real encrypted checkout/inbox list-route integration tests with negative cross-scope assertions, and engine-level warning tests for zero-token and joined-field fallbacks.

**Comment by pkarw** [comment]
Label set to `changes-requested` because review found actionable correctness and coverage issues.

**Comment by pkarw** [comment]
Priority set to `priority-high` because this fix touches encryption, tenant scoping, payment data, and event reliability.

**Comment by pkarw** [comment]
Risk set to `risk-high` because the diff changes shared query contracts and encryption-sensitive, tenant-scoped API behavior.

**Comment by pkarw** [comment]
Added `needs-qa` because the API behavior changes on encryption- and payment-sensitive paths and does not qualify for the automated-verification exemption.

**Comment by pkarw** [comment]
Thanks @marcinwadon — review found actionable items, so I am handing this PR back to you for the next pass. I am also carrying the branch forward through the automated fix loop; the replacement PR will preserve your authorship and credit.

**Comment by pkarw** [comment]
Closing in favor of #4784 (https://github.com/open-mercato/open-mercato/pull/4784). Credit to @marcinwadon for the original implementation. The replacement PR carries the same work forward with the requested fixes so it can merge without waiting on the fork branch.

**Comment by pkarw** [comment]
🤖 `om-auto-review-pr` completed: CHANGES REQUESTED, resolved by 1 autofix iteration in replacement PR #4784. Original lock released.
